casper/src/main/scala/coop/rchain/casper/blocks/proposer/BlockCreator.scala

```diff
@@ -1,228 +1,109 @@
-   *
-   *  1. Rank each of the block cs's latest messages (blocks) via the LMD GHOST estimator.
-   *  2. Let each latest message have a score of 2^(-i) where i is the index of that latest message in the ranking.
-   *     Take a subset S of the latest messages such that the sum of scores is the greatest and
-   *     none of the blocks in S conflicts with each other. S will become the parents of the
-   *     about-to-be-created block.
-   *  3. Extract all valid deploys that aren't already in all ancestors of S (the parents).
-   *  4. Create a new block that contains the deploys from the previous step.
-   */
-  def create[F[_]: Concurrent: Time: RuntimeManager: BlockDagStorage: BlockStore: Log: Metrics: Span](
-      preState: ParentsMergedState,
-      validatorIdentity: ValidatorIdentity,
-      shardId: String,
-      dummyDeployOpt: Option[(PrivateKey, String)] = None
-  ): F[BlockCreatorResult] =
-    Span[F].trace(ProcessDeploysAndCreateBlockMetricsSource) {
-      val selfId         = ByteString.copyFrom(validatorIdentity.publicKey.bytes)
-      val nextSeqNum     = preState.maxSeqNums.get(selfId).map(_ + 1L).getOrElse(0L)
-      val nextBlockNum   = preState.maxBlockNum + 1
-      val justifications = preState.justifications.map(_.blockHash).toList
+final case class BlockCreator(id: ValidatorIdentity, shardId: String) {
+  type StateTransitionResult = (StateHash, Seq[ProcessedDeploy], Seq[ProcessedSystemDeploy])
 
-      def prepareUserDeploys(blockNumber: Long): F[Set[Signed[DeployData]]] =
-        for {
-          unfinalized         <- BlockDagStorage[F].pooledDeploys.map(_.values.toSet)
-          earliestBlockNumber = blockNumber - MultiParentCasper.deployLifespan
-          valid = unfinalized.filter { d =>
-            notFutureDeploy(blockNumber, d.data) &&
-            notExpiredDeploy(earliestBlockNumber, d.data)
-          }
-          // this is required to prevent resending the same deploy several times by validator
-//          validUnique = valid -- s.deploysInScope
-          // TODO: temp solution to filter duplicated deploys
-          validUnique <- valid.toList.filterA { d =>
-                          BlockDagStorage[F].lookupByDeployId(d.sig).map(_.isEmpty)
-                        }
-        } yield validUnique.toSet
+  def create[F[_]: Concurrent: RuntimeManager: BlockDagStorage: BlockStore: Log: Metrics: Span](
+      preStateHash: Blake2b256Hash,
+      parents: Set[BlockHash],
+      bondsMap: Map[Validator, Long],
+      finalization: Set[DeployId], // deploys that are rejected on finalization done by the block being created
+      blockNum: Long,
+      seqNum: Long,
+      deploys: Seq[DeployId],
+      toSlash: Set[Validator],
+      changeEpoch: Boolean,
+      suppressAttestation: Boolean
+  ): F[BlockCreatorResult] = {
+    val creatorsPk    = id.publicKey
+    val blockData     = BlockData(blockNum, creatorsPk, seqNum)
+    val shouldPropose = deploys.nonEmpty || toSlash.nonEmpty || changeEpoch
```

Comment:
We need to split this logic in multiple functions. First to detect conditions and data for the next block and after that the block creation itself.
This is also important to correct handling of invalid blocks (consensus related and replay related) and also for easier testing.
**Create** function should ALWAYS create a block.

Reply:
It looks to me that you want this function to be just "packageBlock".
If we go to deep refactoring, I think that there should not be even `create block` function which is special for propose. It should be the same thing for propose and replaying block.

Reply:
`shouldPropose` can be calculated by the caller so that logic of checking for conditions is not mixed with the main logic of creation blocks.

We can leave it as is in this PR and fix it later.